File: core/blocks/schema.py
"""
Version 1 of the Yana content format on the wire.

Everything about this encoding is explicit, because the alternative -- Swift's
synthesized ``Codable`` output -- ties the format to compiler-internal shapes
(``{"paragraph": {"_0": [...]}}``) and an integer style bitmask that no
non-Swift client can read. Here every block carries a ``type`` discriminator,
``styles`` is a string array, and the version sits once on the envelope.

Two rules make the format extensible, and both are load-bearing on **both**
sides of the wire:

* an unknown block ``type`` is skipped, never fatal;
* an unknown style name is ignored, never fatal.

Optional values are ``null`` on the wire and ``""`` in the dataclasses.
"""

from typing import Any

from .types import (
    EMBED_PROVIDERS,
    FORMAT_VERSION,
    STYLE_NAMES,
    Block,
    Blockquote,
    CodeBlock,
    Divider,
    EmbedBlock,
    Heading,
    ImageBlock,
    InlineRun,
    ListBlock,
    Paragraph,
)


class UnsupportedFormatVersion(ValueError):
    """The envelope's ``version`` is not one this build understands."""


def _or_none(value: str) -> str | None:
    return value or None


def encode_run(run: InlineRun) -> dict[str, Any]:
    return {"text": run.text, "styles": run.styles, "link": _or_none(run.link)}


def encode_block(block: Block) -> dict[str, Any]:
    match block:
        case Paragraph(runs=runs):
            return {"type": "paragraph", "runs": [encode_run(run) for run in runs]}
        case Heading(level=level, runs=runs):
            return {"type": "heading", "level": level, "runs": [encode_run(run) for run in runs]}
        case ListBlock(ordered=ordered, items=items):
            return {
                "type": "list",
                "ordered": ordered,
                "items": [[encode_block(inner) for inner in item] for item in items],
            }
        case Blockquote(blocks=blocks):
            return {"type": "blockquote", "blocks": [encode_block(inner) for inner in blocks]}
        case ImageBlock(ref=ref, caption=caption):
            return {"type": "image", "ref": ref, "caption": [encode_run(run) for run in caption]}
        case EmbedBlock(
            provider=provider, external_url=external_url, thumbnail_ref=thumbnail, title=title
        ):
            return {
                "type": "embed",
                "provider": provider,
                "thumbnailRef": _or_none(thumbnail),
                "externalURL": external_url,
                "title": _or_none(title),
            }
        case CodeBlock(text=text, language=language):
            return {"type": "codeBlock", "text": text, "language": _or_none(language)}
        case Divider():
            return {"type": "divider"}
    raise TypeError(f"not a block: {block!r}")


def encode_document(blocks: list[Block]) -> dict[str, Any]:
    """The full envelope: a version and a block array."""
    return {"version": FORMAT_VERSION, "blocks": [encode_block(block) for block in blocks]}


def _clamp_level(value: Any) -> int:
    try:
        level = int(value)
    except (TypeError, ValueError):
        return 1
    return min(max(level, 1), 6)


def decode_runs(items: Any) -> list[InlineRun]:
    runs: list[InlineRun] = []
    for item in items or []:
        if not isinstance(item, dict):
            continue
        raw = item.get("styles") or []
        styles = {name for name in raw if name in STYLE_NAMES}
        runs.append(
            InlineRun(
                text=item.get("text") or "",
                bold="bold" in styles,
                italic="italic" in styles,
                code="code" in styles,
                strikethrough="strikethrough" in styles,
                link=item.get("link") or "",
            )
        )
    return runs


def decode_block(obj: Any) -> Block | None:
    """Decode one block, or ``None`` for anything unrecognized."""
    if not isinstance(obj, dict):
        return None
    match obj.get("type"):
        case "paragraph":
            return Paragraph(runs=decode_runs(obj.get("runs")))
        case "heading":
            return Heading(
                level=_clamp_level(obj.get("level", 1)), runs=decode_runs(obj.get("runs"))
            )
        case "list":
            return ListBlock(
                ordered=bool(obj.get("ordered", False)),
                items=[decode_blocks(item) for item in (obj.get("items") or [])],
            )
        case "blockquote":
            return Blockquote(blocks=decode_blocks(obj.get("blocks")))
        case "image":
            return ImageBlock(ref=obj.get("ref") or "", caption=decode_runs(obj.get("caption")))
        case "embed":
            provider = obj.get("provider") or "generic"
            return EmbedBlock(
                provider=provider if provider in EMBED_PROVIDERS else "generic",
                external_url=obj.get("externalURL") or "",
                thumbnail_ref=obj.get("thumbnailRef") or "",
                title=obj.get("title") or "",
            )
        case "codeBlock":
            return CodeBlock(text=obj.get("text") or "", language=obj.get("language") or "")
        case "divider":
            return Divider()
    return None


def decode_blocks(items: Any) -> list[Block]:
    """Decode a block array, dropping anything unrecognized."""
    decoded = (decode_block(item) for item in items or [])
    return [block for block in decoded if block is not None]


def decode_document(payload: dict[str, Any]) -> list[Block]:
    version = payload.get("version")
    if version != FORMAT_VERSION:
        raise UnsupportedFormatVersion(f"unsupported content format version: {version!r}")
    return decode_blocks(payload.get("blocks"))

File: src/lib/aggregators/blocks/schema.ts
/**
 * Version 1 of the Yana content format on the wire.
 *
 * Wire format discriminator field is `type`.
 * Extensibility rules:
 * - unknown block type is skipped, never fatal;
 * - unknown style name is ignored, never fatal.
 *
 * Optional values are null on the wire and "" in memory Block objects.
 */

import {
  clampHeadingLevel,
  EMBED_PROVIDERS,
  FORMAT_VERSION,
  STYLE_NAMES,
  type Block,
  type EmbedProvider,
  type InlineRun,
  type StyleName,
} from "./types";

export class UnsupportedFormatVersion extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UnsupportedFormatVersion";
  }
}

export interface WireInlineRun {
  text: string;
  styles: StyleName[];
  link: string | null;
}

export type WireBlock =
  | { type: "paragraph"; runs: WireInlineRun[] }
  | { type: "heading"; level: number; runs: WireInlineRun[] }
  | { type: "list"; ordered: boolean; items: WireBlock[][] }
  | { type: "blockquote"; blocks: WireBlock[] }
  | { type: "summary"; blocks: WireBlock[] }
  | { type: "image"; ref: string; caption: WireInlineRun[] }
  | {
      type: "embed";
      provider: string;
      thumbnailRef: string | null;
      externalURL: string;
      title: string | null;
    }
  | { type: "codeBlock"; text: string; language: string | null }
  | { type: "divider" };

export interface WireDocument {
  version: typeof FORMAT_VERSION;
  blocks: WireBlock[];
}

function orNull(value: string | undefined | null): string | null {
  return value && value.length > 0 ? value : null;
}

export function encodeRun(run: InlineRun): WireInlineRun {
  const styles: StyleName[] = [];
  if (run.bold) styles.push("bold");
  if (run.italic) styles.push("italic");
  if (run.code) styles.push("code");
  if (run.strikethrough) styles.push("strikethrough");

  return {
    text: run.text,
    styles,
    link: orNull(run.link),
  };
}

export function encodeBlock(block: Block): WireBlock {
  switch (block.kind) {
    case "paragraph":
      return {
        type: "paragraph",
        runs: block.runs.map(encodeRun),
      };
    case "heading":
      return {
        type: "heading",
        level: block.level,
        runs: block.runs.map(encodeRun),
      };
    case "list":
      return {
        type: "list",
        ordered: block.ordered,
        items: block.items.map((item) => item.map(encodeBlock)),
      };
    case "blockquote":
      return {
        type: "blockquote",
        blocks: block.blocks.map(encodeBlock),
      };
    case "summary":
      return {
        type: "summary",
        blocks: block.blocks.map(encodeBlock),
      };
    case "image":
      return {
        type: "image",
        ref: block.ref,
        caption: block.caption.map(encodeRun),
      };
    case "embed":
      return {
        type: "embed",
        provider: block.provider,
        thumbnailRef: orNull(block.thumbnailRef),
        externalURL: block.externalUrl,
        title: orNull(block.title),
      };
    case "code_block":
      return {
        type: "codeBlock",
        text: block.text,
        language: orNull(block.language),
      };
    case "divider":
      return {
        type: "divider",
      };
    default: {
      const _exhaustive: never = block;
      throw new TypeError(`not a block: ${JSON.stringify(_exhaustive)}`);
    }
  }
}

export function encodeDocument(blocks: Block[]): WireDocument {
  return {
    version: FORMAT_VERSION,
    blocks: blocks.map(encodeBlock),
  };
}

/**
 * Coerce an untrusted wire value into a heading level. Only the coercion --
 * float truncation, string parsing, NaN defaulting to 1 -- is this function's
 * own concern; the 1-6 bound itself is `clampHeadingLevel()` (`./types`), the
 * one place that arithmetic lives, so this doesn't hold its own copy of it.
 */
function clampLevel(value: unknown): number {
  let level = typeof value === "number" ? Math.floor(value) : parseInt(String(value), 10);
  if (isNaN(level)) level = 1;
  return clampHeadingLevel(level);
}

export function decodeRuns(items: unknown): InlineRun[] {
  if (!Array.isArray(items)) return [];
  const runs: InlineRun[] = [];
  for (const item of items) {
    if (typeof item !== "object" || item === null) continue;
    const raw = item as Record<string, unknown>;
    const rawStyles = Array.isArray(raw.styles) ? raw.styles : [];
    const validStyles = new Set<string>(
      rawStyles.filter(
        (s): s is string => typeof s === "string" && (STYLE_NAMES as readonly string[]).includes(s),
      ),
    );
    const text = typeof raw.text === "string" ? raw.text : "";
    const link = typeof raw.link === "string" ? raw.link : "";

    runs.push({
      text,
      bold: validStyles.has("bold"),
      italic: validStyles.has("italic"),
      code: validStyles.has("code"),
      strikethrough: validStyles.has("strikethrough"),
      link,
    });
  }
  return runs;
}

export function decodeBlock(obj: unknown): Block | null {
  if (typeof obj !== "object" || obj === null) return null;
  const raw = obj as Record<string, unknown>;

  switch (raw.type) {
    case "paragraph":
      return {
        kind: "paragraph",
        runs: decodeRuns(raw.runs),
      };
    case "heading":
      return {
        kind: "heading",
        level: clampLevel(raw.level ?? 1),
        runs: decodeRuns(raw.runs),
      };
    case "list": {
      const rawItems = Array.isArray(raw.items) ? raw.items : [];
      const items: Block[][] = rawItems.map((item) => decodeBlocks(item));
      return {
        kind: "list",
        ordered: Boolean(raw.ordered),
        items,
      };
    }
    case "blockquote":
      return {
        kind: "blockquote",
        blocks: decodeBlocks(raw.blocks),
      };
    case "summary":
      return {
        kind: "summary",
        blocks: decodeBlocks(raw.blocks),
      };
    case "image":
      return {
        kind: "image",
        ref: typeof raw.ref === "string" ? raw.ref : "",
        caption: decodeRuns(raw.caption),
      };
    case "embed": {
      const rawProvider = typeof raw.provider === "string" ? raw.provider : "generic";
      const provider = (EMBED_PROVIDERS as readonly string[]).includes(rawProvider)
        ? (rawProvider as EmbedProvider)
        : "generic";
      return {
        kind: "embed",
        provider,
        externalUrl: typeof raw.externalURL === "string" ? raw.externalURL : "",
        thumbnailRef: typeof raw.thumbnailRef === "string" ? raw.thumbnailRef : "",
        title: typeof raw.title === "string" ? raw.title : "",
      };
    }
    case "codeBlock":
      return {
        kind: "code_block",
        text: typeof raw.text === "string" ? raw.text : "",
        language: typeof raw.language === "string" ? raw.language : "",
      };
    case "divider":
      return {
        kind: "divider",
      };
    default:
      return null;
  }
}

export function decodeBlocks(items: unknown): Block[] {
  if (!Array.isArray(items)) return [];
  const blocks: Block[] = [];
  for (const item of items) {
    const decoded = decodeBlock(item);
    if (decoded !== null) {
      blocks.push(decoded);
    }
  }
  return blocks;
}

export function decodeDocument(payload: unknown): Block[] {
  if (typeof payload !== "object" || payload === null) {
    throw new UnsupportedFormatVersion(
      `unsupported content format version: ${JSON.stringify(payload)}`,
    );
  }
  const raw = payload as Record<string, unknown>;
  if (raw.version !== FORMAT_VERSION) {
    throw new UnsupportedFormatVersion(
      `unsupported content format version: ${JSON.stringify(raw.version)}`,
    );
  }
  return decodeBlocks(raw.blocks);
}
